Add more stuff/options to build (#23)
godotengine/godot-cpp-template: README.md

* GitHub CI/CD to publish your library packages when creating a release (`.github/workflows/builds.yml`)
* preconfigured source files for C++ development of the GDExtension (`src/`)

## Usage - Actions

The actions builds `godot-cpp` at a specified location, and then builds the `gdextension` at a configurable location. It builds for desktop, mobile and web and allows for configuration on what platforms you need. It also supports configuration for debug and release builds, and for double builds.

> The action uses SConstruct for both godot-cpp and gdextension that is built.

Reviewer: ```suggestion
The action uses SConstruct for both godot-cpp and the GDExtension that is built.
```

          - platform: android
            arch: arm64
            os: ubuntu-20.04
          - platform: android
            arch: arm32
            os: ubuntu-20.04
          - platform: android
            arch: x86_64
            os: ubuntu-20.04
          - platform: android
            arch: x86_32
            os: ubuntu-20.04
          - platform: ios
            arch: arm64
            os: macos-latest
          - platform: web
            arch: wasm32
            os: ubuntu-20.04

    runs-on: ${{ matrix.os }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          submodules: true
          fetch-depth: 0
      - name: 🔗 GDExtension Build
        uses: ughuuu/godot-cpp-template/.github/actions/build@main
        with:
          platform: ${{ matrix.platform }}
          arch: ${{ matrix.arch }}
          godot-cpp-location: godot-cpp
          float-precision: single
          build-target-type: template_release
      - name: Upload Artifact
        uses: actions/upload-artifact@v3
        with:
          name: GDExtension
          path: |
            ${{ github.workspace }}/bin/**

```

## Usage - Signing

You will need:

> - A macbook
> - An apple id enrolled in apple developer(99 dollar per year)
> - In the framework folder, you need to create a `Resources/Info.plist`. Take as example the one in this project. Be careful to set CFBundleExecutable to the EXACT lib name, otherwise it won't work. Also, don't put strange names in the CFBundleName and other such places, try to only use letters and space. Errors will be if not extremly vague to impossible to debug.
> 
> For the actions you will need to set the following inputs. Store them as secrets in github:

Reviewer: ```suggestion
- A Mac
- An Apple ID enrolled in Apple Developer Program (99 USD per year)
- A `Resources/Info.plist` in the `framework` folder. Take the one in this project as an example. Be careful to set CFBundleExecutable to the **EXACT** lib name, otherwise it won't work. Also, don't put strange names in the CFBundleName and other such places. Try to only use letters and spaces. Errors will be extremly vague if not impossible to debug.

For the actions you will need to set the following inputs. Store them as secrets in GitHub:
```

- APPLE_CERT_BASE64
- APPLE_CERT_PASSWORD
- APPLE_DEV_ID
- APPLE_DEV_TEAM_ID
- APPLE_DEV_PASSWORD
- APPLE_DEV_APP_ID

> You will find here a guide on how to create all of them. Go to developer.apple.com:

Reviewer: ```suggestion
You will find here a guide on how to create all of them. Go to [developer.apple.com](developer.apple.com):
```

> - Create an Apple ID if you don’t have one already.
> - Use your Apple ID to register as an Apple Developer.
> - Accept all agreements from Apple Developer Page.

Reviewer: ```suggestion
- Create an Apple ID if you don’t have one already.
- Use your Apple ID to register in the Apple Developer Program.
- Accept all agreements from the Apple Developer Page.
```

> ### APPLE_DEV_ID - Apple Id
> 
> - Your email used for your apple id.

Reviewer: ```suggestion
### APPLE_DEV_ID - Apple ID

- Your email used for your Apple ID.
```

- APPLE_DEV_ID = email@provider.com

> ### APPLE_DEV_TEAM_ID - Apple Team Id
> 
> - Go to [developer.apple.com](https://developer.apple.com). Go to account.
> - Go to membership details. Copy Team Id.

Reviewer: ```suggestion
### APPLE_DEV_TEAM_ID - Apple Team ID

- Go to [developer.apple.com](https://developer.apple.com). Go to account.
- Go to membership details. Copy Team ID.
```

- APPLE_DEV_TEAM_ID = `1ABCD23EFG`

### APPLE_DEV_PASSWORD - Apple App-Specific Password

- Create [Apple App-Specific Password](https://support.apple.com/en-us/102654). Copy the password.

- APPLE_DEV_PASSWORD = `abcd-abcd-abcd-abcd`

### APPLE_CERT_BASE64 and APPLE_CERT_PASSWORD and APPLE_DEV_APP_ID

- Go to [developer.apple.com](https://developer.apple.com). Go to account.
- Go to certificates.
- Click on + at Certificates tab. Create Developer ID Application. Click Continue.
- Leave profile type as is. [Create a certificate signing request from a mac](https://developer.apple.com/help/account/create-certificates/create-a-certificate-signing-request). You can use your own name and email address. Save the file to disk. You will get a file called `CertificateSigningRequest.certSigningRequest`. Upload it to the Developer ID Application request. Click Continue.
- Download the certificate. You will get a file `developerID_application.cer`.

> - On a macbook, right click and select open. Add it to the login keychain. In the Keychain Access app that opened, login Keychain tab, go to Keys, sort by date modified, expand your key (the key should have name you entered at common name `Common Name`), right click the expanded certificate, get info, and copy the text at Details -> Subject Name -> Common Name.
> 

Reviewer: ```suggestion
- On a Mac, right click and select open. Add it to the login keychain. In the Keychain Access app that opened, login Keychain tab, go to Keys, sort by date modified, expand your key (the key should have name you entered at common name `Common Name`), right click the expanded certificate, get info, and copy the text at Details -> Subject Name -> Common Name.

```

Eg.
- APPLE_DEV_APP_ID = `Developer ID Application: Common Name (1ABCD23EFG)`

- Then, select the certificate, right click and click export. At file format select p12. When exporting, set a password for the certificate. This will be APPLE_CERT_PASSWORD. You will get a `Certificates.p12` file.

Eg.
- APPLE_CERT_PASSWORD = `<password_set_when_exporting_p12>`

- Then you need to make a base64 file out of it, by running:
```
base64 -i Certificates.p12 -o Certificates.base64
```

- Copy the contents of the generated file:
Eg.
- APPLE_CERT_BASE64 = `...`(A long text file)

- While still

## Usage - Template

> To use this template, log in to github and click the green "Use this template" button at the top of the repository page.

Reviewer: ```suggestion
To use this template, log in to GitHub and click the green "Use this template" button at the top of the repository page.
```
